File: Problem Sets/PSet 1/Credit/credit.c
/* 
:: Valid creditcard numbers for test ::

AMEX
    378282246310005 
MASTERCARD
    5105105105105100
VISA
    4012888888881881
*/

#include <stdio.h>
#include <cs50.h>

int main()
{
    long long number = get_long_long("What's your credit card number? "); // requesting entry of the card number
    int sumdouble = 0, sumevens = 0, numdigits = 0, n1 = 0, n2 = 0;

    while (number > 0)
    {
        n2 = n1;
        n1 = number % 10;

        if (numdigits % 2 != 0)
        {
            int multiple = 2 * n1;
            sumdouble += (multiple / 10) + (multiple % 10);   
        }
        else
        {
            sumevens += n1;
        }
        number /= 10;
        numdigits++;
    }
    bool valid = (sumdouble + sumevens) % 10 == 0;
    int f_two_digits = (n1 * 10) + n2;

    // Valdation of flags
    // VISA credit card validation
    if (n1 == 4 && numdigits >= 13 && numdigits <= 16 && valid)
    {
        printf("VISA\n");
    }
    // MasterCard credit card validation
    else if (f_two_digits >= 51 && f_two_digits <= 55 && numdigits == 16 && valid)
    {
        printf("MASTERCARD\n");
    }
    // AMEX credit card validation
    else if ((f_two_digits == 34 || f_two_digits == 37) && numdigits == 15 && valid)
    {
        printf("AMEX\n");
    }
    // if the user enter a invalid number
    else 
    {
        printf("Sorry, number is INVALID\n");
    }
}

File: Problem Sets/PSet 6/Credit/credit.py
'''
:: Valid creditcard numbers for test ::

AMEX
    378282246310005
MASTERCARD
    5105105105105100
VISA
    4012888888881881
'''

from cs50 import get_int
# Number of cc
while True:
    card = get_int("What's your credit card number? ")
    # if the user enters with an character
    if card >= 0:
        break

firstDigits = 0
first_two_Digits = 0
digits = 0
numbers = 0

while card >= 1:
    rest = card % 10
    
    if card >= 10 and card < 100:
        first_two_Digits = card
    
    if card < 10:
        firstDigits = card
        
    if digits % 2 == 1:
        num = rest * 2
        
        if num > 9:
            num = num // 10 + num % 10
        
        numbers += num
    else:
        numbers += rest
        
    card //= 10
    
    digits += 1
    
# Validation cc
if numbers % 10 == 0:
    # Validation AMEX
    if digits == 15 and (first_two_Digits == 34 or first_two_Digits == 37):
        print("AMEX")
    # Validation MasterCard
    elif digits == 16 and first_two_Digits >= 51 and first_two_Digits <= 55:
        print("MASTERCARD")
    # Validation VISA
    elif (digits == 13 or digits == 16) and firstDigits == 4:
        print("VISA")
    else:
        # if the user enters an invalid number
        print("INVALID")
# if the user enters with an different numbers from the initial digits of the card
else:
    print("INVALID")